mlx_graphs/data/data.py

```diff
@@ -0,0 +1,61 @@
+from typing import Optional
+
+import mlx.core as mx
+import mlx.data as dx
+
+
+class Data:
+    """
+    Represents a graph data object with optional features and labels.
+
+    Args:
+        edge_index (mlx.core.array, optional): edge index representing the topology of
+            the graph, with shape [2, num_edges].
+        node_features (mlx.core.array, optional): Array of shape [num_nodes, num_node_features]
+            containing the features of each node.
+        edge_features (mlx.core.array, optional): Array of shape [num_edges, num_edge_features]
+            containing the features of each edge.
+        graph_features (mlx.core.array, optional): Array of shape [num_graph_features]
+            containing the global features of the graph.
+        node_labels (mlx.core.array, optional): Array of shape [num_nodes, num_node_labels]
+            containing the labels of each node.
+        edge_labels (mlx.core.array, optional): Array of shape [num_edges, num_edge_labels]
+            containing the labels of each edge.
+        graph_labels (mlx.core.array, optional): Array of shape [num_graph_labels]
+            containing the global labels of the graph.
+        **kwargs: Additional keyword arguments to store any other custom attributes.
+    """
+
+    def __init__(
+        self,
+        edge_index: Optional[mx.array] = None,
+        node_features: Optional[mx.array] = None,
+        edge_features: Optional[mx.array] = None,
+        graph_features: Optional[mx.array] = None,
+        node_labels: Optional[mx.array] = None,
+        edge_labels: Optional[mx.array] = None,
+        graph_labels: Optional[mx.array] = None,
+        **kwargs,
+    ):
+        self.edge_index = edge_index
+        self.node_features = node_features
+        self.edge_features = edge_features
+        self.graph_features = graph_features
+        self.node_labels = node_labels
+        self.edge_labels = edge_labels
+        self.graph_labels = graph_labels
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    def to_dict(self):
+        """Converts the Data object to a dictionary.
+
+        Returns:
+            dict: A dictionary representation of the Data object.
+        """
+        return self.__dict__
+
+
+class Dataset:
+    def __init__(self, data: list[Data]) -> None:
+        self.dataset = dx.buffer_from_vector(data)  # type: ignore
```

Comment:
Did you find documentation about `buffer_from_vector`?
I'm not sure this allows to load `Data` lazily one by one.

Reply:
It looks like `Data` has to be iterable, got this error: `'Data' object is not iterable
`

Reply:
Just found the doc [here](https://ml-explore.github.io/mlx-data/build/html/python/_autosummary/mlx.data.buffer_from_vector.html#mlx.data.buffer_from_vector).

Reply:
Yes, that was WIP - should be ok now.

However, I'm testing it and
```
self.dataset = dx.buffer_from_vector([x.to_dict() for x in data])
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: contiguous array expected -- use numpy.ascontiguousarray()
```

Reply:
> Did you find documentation about `buffer_from_vector`? I'm not sure this allows to load `Data` lazily one by one.

On this - from the docs it seems that Data should be loaded lazily https://ml-explore.github.io/mlx-data/build/html/python/_autosummary/mlx.data.buffer_from_vector.html

Reply:
Ok then we have to provide 2 options:
* load a dataset from a list of `Data`, this will be not lazy as the list of `Data` is already in memory.
* load a dataset from a list of files that store `Data` objects, this will loads each `Data` lazily from disk.

Reply:
yes, that's an option. The only issue I see with that is that storing `Data` objects as files may not be optimal. Perhaps there's a way to automatically do that when you pass a list of `Data`.

Reply:
BTW: it seems that `mlx.data` requires to work with `numpy` arrays instead of `mlx` arrays.

Reply:
> BTW: it seems that `mlx.data` requires to work with `numpy` arrays instead of `mlx` arrays.

Hmm ok strange

Reply:
> yes, that's an option. The only issue I see with that is that storing `Data` objects as files may not be optimal. Perhaps there's a way to automatically do that when you pass a list of `Data`.

I'm pretty sure that it's the only solution for real lazy loading. Take this example: you have a dataset with 10 graphs of size 10GB. Initializing `Dataset` with the 10 `Data` is not feasible. Instead, we have either to load each `Data` directly from a kind of pkl file or to download it from internet and then load the `Data`.

Reply:
> > yes, that's an option. The only issue I see with that is that storing `Data` objects as files may not be optimal. Perhaps there's a way to automatically do that when you pass a list of `Data`.
> 
> I'm pretty sure that it's the only solution for real lazy loading. Take this example: you have a dataset with 10 graphs of size 10GB. Initializing `Dataset` with the 10 `Data` is not feasible. Instead, we have either to load each `Data` directly from a kind of pkl file or to download it from internet and then load the `Data`.

Yes, sure, in that case it makes sense - but perhaps we should have two different classes for the two cases?

Reply:
> > BTW: it seems that `mlx.data` requires to work with `numpy` arrays instead of `mlx` arrays.
> 
> Hmm ok strange

This works
```
dx.buffer_from_vector([{"x": np.zeros([1, 1])} for i in range(10)])
```
This raises a `ValueError`
```
dx.buffer_from_vector([{"x": mx.zeros([1, 1])} for i in range(10)])
```

I'll ask in `mlx-data` whether that's desired behavior.

Reply:
> Yes, sure, in that case it makes sense - but perhaps we should have two different classes for the two cases?

Maybe

Reply:
> > > BTW: it seems that `mlx.data` requires to work with `numpy` arrays instead of `mlx` arrays.
> > 
> > 
> > Hmm ok strange
> 
> This works
> 
> ```
> dx.buffer_from_vector([{"x": np.zeros([1, 1])} for i in range(10)])
> ```
> 
> This raises a `ValueError`
> 
> ```
> dx.buffer_from_vector([{"x": mx.zeros([1, 1])} for i in range(10)])
> ```
> 
> I'll ask in `mlx-data` whether that's desired behavior.

Answer should be [here](https://ml-explore.github.io/mlx-data/build/html/buffers_streams_samples.html#samples)

Reply:
We'll see if it's a good choice to use mlx-data. It seems that the advantages of using this package are mainly for image/text processing. If it just provides a wrapper class with an iterator on the data, I don't think it's worth it to add this dependency as we can create a dedicated class for that. However, it seems that it is fast to load objects from files (as shown in the example with images), so maybe this would be something to explore.

Reply:
Ok, shall we remove it for now and perhaps reintroduce it later if needed?

Reply:
Yes we can do that. We can give it another try when implementing the lazy `Dataset` class for large graphs.

Reply:
Let me leave just the `Data` class in this PR so that we can start using that one at least.

Reply:
@Kheims can you submit a PR with your beginning work on batching? So that we can have a look on what to do next.

Reply:
> Let me leave just the `Data` class in this PR so that we can start using that one at least.

Ok! So you remove the mlx-data stuff?

Reply:
Yes, now it contains just the `Data` class, very minimal PR but we need that block anyways.